libs/s25main/ingameWindows/iwTools.cpp

```diff
@@ -56,9 +54,9 @@ iwTools::iwTools(const GameWorldViewer& gwv, GameCommandFactory& gcFactory)
         for(unsigned i = 0; i < NUM_TOOLS; ++i)
         {
             Extent btSize = Extent(20, 13);
-            ctrlButton* bt = AddImageButton(100 + i * 2, DrawPoint(174, 25 + i * 28), btSize, TC_GREY, LOADER.GetImageN("io", 33), "+1");
+            ctrlButton* bt = AddImageButton(100 + i * 2, DrawPoint(174, 27 + i * 28), btSize, TC_GREY, LOADER.GetImageN("io", 33), "+1");
             AddImageButton(101 + i * 2, bt->GetPos() + DrawPoint(0, btSize.y), btSize, TC_GREY, LOADER.GetImageN("io", 34), "-1");
-            AddTextDeepening(200 + i, DrawPoint(151, 4 + bt->GetPos().y), Extent(20, 18), TC_GREY, "", NormalFont, COLOR_YELLOW);
```

Comment:
is it possible to get the size from the font itself?

Reply:
Yes and no. There is `font.getDx()` which gets the largest width. However we'd have to add the border the element adds, so we'd need to query the element. I guess the only viable solution would be to add a function `extendWidthForNChars(numberOfChars)` to those classes to make it handle it itself.

Reply:
@Flow86 There is a problem: The value used is to big. The max width is 12 for all chars but the max width for the numbers is 7. So the field gets too big. How'd you handle this? 

My idea: A a function to query the min size of the control and a free function to get the max size for a given char array. Then one can do: `txt->Resize(text->getbordersize() + Extent(2 * getMaxSize(font, {'0', '1', ... '9'}), ...))`

That might look strange but works. Also for the AI button it would be `bt->Resize(bt->getbordersize() + font->getextend(bt->gettext()))`

Reply:
yes I thought we have some "how big is it in real"-method in the font-draw stuff

Reply:
Yes we do. But a) currently missing is a "how much space is there for the text" part and b) something to calculate the max size for a given set of possible texts.

Asking the font "get me the size for '0' to '99'" isn't feasible, but getting the max char width of '0'-'9' and multiplying that by 2 is ok.

There are also 2 design issues:
- Currently the text is part of the textdeepening or textbutton. But IMO it should rather be a child of that. Then it would just be a ctor overload.
- This means that a ctrlText should be sized too. This would also make more sense for calling the Font.Draw method with the alignment: It should pass a rect instead of a point. Avoids stuff like this: https://github.com/Return-To-The-Roots/s25client/blob/d122e7e3f19b86196d02407ebedd485016ec3acf/libs/s25main/controls/ctrlTextDeepening.cpp#L40-L56

If we make ctrlText a (sized) child, then we can generically get the border size by calling `bt->getSize() - bt->getCtrl<ctrlText>(0)->getSize()`

Reply:
Or even better: Make the Font.Draw way more low-level: Let it simply draw the given text at a given position. The alignment and trimming to a size can be done by ctrlText. So everywhere you want that behavior you got to use a ctrlText.